File: example_usb_camera.py
#!/usr/bin/env python3
"""End-to-end example: publish a USB camera into Cyclops.

    USB camera -> OpenCV -> CyclopsCameraPublisher -> eCAL S1/cama -> Cyclops

Swap `cv2.VideoCapture` for your camera SDK; the publishing call is the
same. Simpler than `/usr/bin/usb_camera_publisher.py`, which also handles
auto-detection, hot-plug, v4l2 introspection, and thermal cameras.

Usage:
    python3 example_usb_camera.py
    python3 example_usb_camera.py --device /dev/video2 --config my_cam.yaml

Dependencies and schema lookup are documented in `cyclops_camera_ecal.py`.

Troubleshooting:
- "Failed to open camera": wrong device, or another process has it.
  Run `v4l2-ctl --list-devices`.
- Cyclops doesn't see frames: confirm the topic with `ecal_mon_gui` or
  `example_image_subscriber.py`. Different eCAL networks won't see each
  other.
- Calibration wrong on subscriber: check you edited the YAML at `--config`,
  not the example.
"""

from __future__ import annotations

import argparse
import logging
import signal
import sys
import time
from pathlib import Path

import cv2
import yaml

from cyclops_camera_ecal import (
    CyclopsCameraPublisher,
    load_camera_calibration,
)

logger = logging.getLogger("example_usb_camera")


def open_camera(device: str, width: int, height: int, fps: int) -> cv2.VideoCapture:
    """Open a USB camera with OpenCV. Replace with your own camera SDK."""
    capture_target: int | str = int(device) if device.lstrip("-").isdigit() else device
    camera = cv2.VideoCapture(capture_target)
    if not camera.isOpened():
        raise RuntimeError(f"Failed to open camera '{device}'")

    camera.set(cv2.CAP_PROP_FRAME_WIDTH, width)
    camera.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
    camera.set(cv2.CAP_PROP_FPS, fps)

    actual_w = int(camera.get(cv2.CAP_PROP_FRAME_WIDTH))
    actual_h = int(camera.get(cv2.CAP_PROP_FRAME_HEIGHT))
    actual_fps = camera.get(cv2.CAP_PROP_FPS)
    logger.info(f"Opened {device}: {actual_w}x{actual_h} @ {actual_fps:.1f} fps")
    return camera


def load_yaml(path: Path) -> dict:
    """Read raw YAML for capture parameters not in `CameraCalibration` (resolution, fps)."""
    with path.open("r", encoding="utf-8") as handle:
        return yaml.safe_load(handle)


def main() -> int:
    parser = argparse.ArgumentParser(description="Publish a USB camera to Cyclops on S1/cama")
    parser.add_argument(
        "--device",
        default="0",
        help="OpenCV camera index (e.g. 0) or device path (e.g. /dev/video0). Default: 0.",
    )
    parser.add_argument(
        "--config",
        type=Path,
        default=Path(__file__).resolve().parent / "camera_params.example.yaml",
        help="Path to camera_params.yaml. Default: camera_params.example.yaml next to this script.",
    )
    parser.add_argument("--log-level", default="INFO", choices=["DEBUG", "INFO", "WARNING", "ERROR"])
    args = parser.parse_args()

    logging.basicConfig(
        level=getattr(logging, args.log_level),
        format="%(asctime)s [%(levelname)s] %(name)s: %(message)s",
    )

    if not args.config.exists():
        logger.error(f"Config not found: {args.config}")
        return 1

    # One YAML feeds both the calibration (capnp Image fields) and the
    # cv2 capture properties — published intrinsics match what's captured.
    calibration = load_camera_calibration(args.config)
    raw_config = load_yaml(args.config)
    resolution = raw_config.get("resolution", {})
    width = int(resolution.get("width", 640))
    height = int(resolution.get("height", 512))
    fps = int(raw_config.get("fps", 30))

    camera = open_camera(args.device, width, height, fps)
    publisher = CyclopsCameraPublisher()

    stop_requested = {"value": False}

    def request_stop(sig, _frame):
        logger.info(f"Signal {sig} received, stopping")
        stop_requested["value"] = True

    signal.signal(signal.SIGINT, request_stop)
    signal.signal(signal.SIGTERM, request_stop)

    frame_interval = 1.0 / max(1, fps)
    consecutive_failures = 0

    try:
        while not stop_requested["value"]:
            t0 = time.time()
            ok, frame = camera.read()
            if not ok or frame is None:
                consecutive_failures += 1
                if consecutive_failures >= 10:
                    logger.error("Camera read failed 10 times in a row; exiting")
                    return 1
                time.sleep(0.05)
                continue
            consecutive_failures = 0

            try:
                publisher.publish_frame(frame, calibration)
            except Exception as exc:
                logger.warning(f"Publish failed: {exc}")

            # Bound publish rate; cv2 often already blocks at native fps.
            elapsed = time.time() - t0
            if elapsed < frame_interval:
                time.sleep(frame_interval - elapsed)
    finally:
        camera.release()
        publisher.close()
        logger.info("Shutdown complete")

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: example_usb_camera.cpp
// End-to-end example: publish a USB camera into Cyclops (C++).
//
//     USB camera -> OpenCV -> CyclopsCameraPublisher -> eCAL S1/cama -> Cyclops
//
// C++ counterpart to example_usb_camera.py. Swap cv::VideoCapture for your
// camera SDK; the publishing call is the same.
//
// Build: cmake -S . -B build && cmake --build build
// Run  : ./build/example_usb_camera [--device /dev/videoN] [--config my_cam.yaml]
//
// Dependencies: eCAL, Cap'n Proto, OpenCV (libopencv-dev), yaml-cpp
// (libyaml-cpp-dev). Schema lookup details live in cyclops_camera_ecal.hpp.

#include "cyclops_camera_ecal.hpp"

#include <yaml-cpp/yaml.h>
#include <opencv2/imgproc.hpp>
#include <opencv2/videoio.hpp>

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdlib>
#include <cstring>
#include <iostream>
#include <stdexcept>
#include <string>
#include <thread>

namespace cam = theseus::cyclops::camera;

namespace {
std::atomic<bool> g_stop_requested{false};

void requestStop(int sig) {
    std::cerr << "\nSignal " << sig << " received, stopping\n";
    g_stop_requested.store(true);
}
}  // namespace

struct CliArgs {
    std::string device = "0";
    std::string config = "camera_params.example.yaml";
};

CliArgs parseArgs(int argc, char** argv) {
    CliArgs args;
    for (int i = 1; i < argc; ++i) {
        std::string flag(argv[i]);
        if ((flag == "--device" || flag == "--config") && i + 1 < argc) {
            const std::string value(argv[++i]);
            if (flag == "--device") args.device = value;
            else if (flag == "--config") args.config = value;
        } else if (flag == "-h" || flag == "--help") {
            std::cout << "Usage: " << argv[0] << " [--device <idx|/dev/videoN>] [--config <yaml>]\n";
            std::exit(0);
        } else {
            std::cerr << "Unknown argument: " << flag << "\n";
            std::exit(2);
        }
    }
    return args;
}

// Open a USB camera with OpenCV. Replace with your own camera SDK.
cv::VideoCapture openCamera(const std::string& device, int width, int height, int fps) {
    cv::VideoCapture camera;
    char* end = nullptr;
    const long parsed = std::strtol(device.c_str(), &end, 10);
    const bool is_index = (end != device.c_str() && *end == '\0');
    if (is_index) {
        camera.open(static_cast<int>(parsed));
    } else {
        camera.open(device);
    }
    if (!camera.isOpened()) {
        throw std::runtime_error("Failed to open camera '" + device + "'");
    }
    camera.set(cv::CAP_PROP_FRAME_WIDTH, width);
    camera.set(cv::CAP_PROP_FRAME_HEIGHT, height);
    camera.set(cv::CAP_PROP_FPS, fps);

    const int actual_w = static_cast<int>(camera.get(cv::CAP_PROP_FRAME_WIDTH));
    const int actual_h = static_cast<int>(camera.get(cv::CAP_PROP_FRAME_HEIGHT));
    const double actual_fps = camera.get(cv::CAP_PROP_FPS);
    std::cerr << "Opened " << device << ": " << actual_w << "x" << actual_h << " @ " << actual_fps << " fps\n";
    return camera;
}

// Load camera_params.yaml. The C++ helper doesn't ship a YAML loader; this
// mirrors the Python helper's load_camera_calibration shape.
cam::CameraCalibration loadCalibration(const std::string& path) {
    YAML::Node root = YAML::LoadFile(path);
    cam::CameraCalibration cal;

    const auto& intr = root["intrinsics"];
    cal.intrinsics.fx = intr["fx"].as<float>();
    cal.intrinsics.fy = intr["fy"].as<float>();
    cal.intrinsics.cx = intr["cx"].as<float>();
    cal.intrinsics.cy = intr["cy"].as<float>();
    cal.intrinsics.k1 = intr["k1"].as<float>(0.0f);
    cal.intrinsics.k2 = intr["k2"].as<float>(0.0f);
    cal.intrinsics.k3 = intr["k3"].as<float>(0.0f);
    cal.intrinsics.k4 = intr["k4"].as<float>(0.0f);

    const auto& body = root["extrinsics"]["body_frame"];
    cal.body_frame.position_x_m = body["position"]["x"].as<double>();
    cal.body_frame.position_y_m = body["position"]["y"].as<double>();
    cal.body_frame.position_z_m = body["position"]["z"].as<double>();
    cal.body_frame.orientation_wxyz = {
        body["orientation"]["w"].as<double>(),
        body["orientation"]["x"].as<double>(),
        body["orientation"]["y"].as<double>(),
        body["orientation"]["z"].as<double>(),
    };

    return cal;
}

int main(int argc, char** argv) {
    const CliArgs args = parseArgs(argc, argv);

    std::signal(SIGINT, requestStop);
    std::signal(SIGTERM, requestStop);

    // One YAML feeds both the calibration and the cv::VideoCapture properties.
    YAML::Node raw = YAML::LoadFile(args.config);
    const int width = raw["resolution"]["width"].as<int>(640);
    const int height = raw["resolution"]["height"].as<int>(512);
    const int fps = raw["fps"].as<int>(30);
    const cam::CameraCalibration calibration = loadCalibration(args.config);

    cv::VideoCapture camera = openCamera(args.device, width, height, fps);
    cam::CyclopsCameraPublisher publisher;

    const auto frame_interval = std::chrono::duration<double>(1.0 / std::max(1, fps));
    int consecutive_failures = 0;
    cv::Mat frame;

    while (!g_stop_requested.load()) {
        const auto t0 = std::chrono::steady_clock::now();

        if (!camera.read(frame) || frame.empty()) {
            if (++consecutive_failures >= 10) {
                std::cerr << "Camera read failed 10 times in a row; exiting\n";
                return 1;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(50));
            continue;
        }
        consecutive_failures = 0;

        try {
            publisher.publishFrame(frame, calibration);
        } catch (const std::exception& exc) {
            std::cerr << "Publish failed: " << exc.what() << "\n";
        }

        // Bound publish rate; cv::VideoCapture often blocks at native fps.
        const auto elapsed = std::chrono::steady_clock::now() - t0;
        if (elapsed < frame_interval) {
            std::this_thread::sleep_for(frame_interval - elapsed);
        }
    }

    camera.release();
    std::cerr << "Shutdown complete\n";
    return 0;
}
